enhancements/installer/cluster-api-based-installations.md

```diff
@@ -0,0 +1,485 @@
+- vSphere: [infrastructure.cluster.x-k8s.io/v1beta1][vsphereCRD]
+
+### Implementation Details/Notes/Constraints [optional]
+
+#### Overview
+
+In a typical CAPI installation, manifests indicating the desired cluster configuration are applied to a
+management cluster. In order to keep `openshift-install` free of new external runtime dependencies,
+the management-cluster dependencies will be [embedded][embed] into the `openshift-install` binary,
+extracted at runtime, and cleaned up afterward.  This approach is similar to what we have been using
+for Terraform.
+
+In its current configuration, the Installer embeds the Terraform and cloud-specific provider binaries
+within the `openshift-install` binary and extracts them at runtime. The Installer produces the Terraform
+configuration files and invokes Terraform using the `tf-exec` library.
+
+![alt terraform diagram](terraform_embedded.jpg)
+
+
+We can follow a similar pattern to run CAPI controllers locally on the Installer host. In addition
+to the CAPI controller binaries, `kube-apiserver` and `etcd` are embedded in order to run a local
+control plane, orchestrated with `envtest`.
+
+![alt capi diagram](capi_embedded.jpg)
+
+#### Local control plane
+
+The local control plane is setup using the previously available work done in Controller Runtime through [envtest][envtest].
+Envtest was born due to a necessity to run integration tests for controllers against a real API server, register webhooks
+(conversion, admission, validation), and managing the lifecycle of Custom Resource Definitions.
+
+Over time, `envtest` matured in a way that now can be used to run controllers in a local environment,
+reducing or eliminating the need for a full Kubernetes cluster to run controllers.
+
+At a high level, the local control plane is responsible for:
+- Setting up certificates for the apiserver and etcd.
+- Running (and cleaning up, on shutdown) the local control plane components.
+- Installing any required component, like Custom Resource Definitions (CRDs)
+    - For Cluster API core the CRDs are stored in `data/data/cluster-api/core-components.yaml`.
+    - Infrastructure providers are expected to store their components in `data/data/cluster-api/<name>-infrastructure-components.yaml`
+- Upon install, the local control plane takes care of modifying any webhook (conversion, admission, validation) to point to the `host:post` combination assigned.
+    - Each controller manager will have its own `host:port` combination assigned.
+    - Certificates are generated and injected in the server, and the client certs in the api-server webhook configuration.
+- For each process that the local control plane manages, a health check (ping to `/healthz`) is required to pass similarly how, when running in a Deployment, a health probe is configured.
+
+#### Manifests
+
+The Installer will produce the CAPI manifests as part of the `manifests` target, writing them to a new
```

Comment:
Making them part of the same target means they can't depend on anything else that's part of the target. So e.g. fields that the user edits in the manifests dir cannot be automatically used to generate changes to the CAPI manifests, instead the user would have to edit in both places.
I suggest we consider adding a different target, because it's difficult to be very, very sure that this situation would never come up.

Reply:
Might be missing something, in which case the user would have to edit in both places?

The Cluster API manifests should only ever be generated (today, at least) when the install-config has the proper feature flag enabled.

Reply:
The manifests that exist today control software level things within the cluster, typically this means users can add additional manifests to configure say, some extra networking or a way to partition disks via additional MachineConfig. One other thing they can do is to configure feature gates.

Do we see any overlap with the infrastructure that will be defined in the openshift manifests? I think supporting detection of user initiated changes may become particularly tricky, and that the overlap between what's inside the cluster and the infrastructure that becomes the cluster is pretty minimal no? The `Infrastructure` object and the CAPI manifests are the most likely to have any overlap IMO.

Reply:
> Might be missing something, in which case the user would have to edit in both places?

In the case where something is configurable via the manifests but not the install-config.
We endeavour to keep the ignition-config as simple as possible, so a few long-tail config options are not exposed there.
It's possible that this wouldn't be an issue, but almost impossible to _prove_ that it won't be an issue.

Reply:
Added to alternatives section. Marking the thread as resolved, but please feel free to reopen if there is more to cover here.

Reply:
Reopening this thread, as I realized @pierreprinetti has brought up a good example of the problem: if users make masters schedulable by changing  `cluster-scheduler-02-config.yml`, then OpenStack needs to add certain security groups.

So the user would need to edit the above manifest and then add the security groups in a CAPI manifest. 

While not ideal, I think that this is fine in terms of UX and is best handled through docs.

Reply:
It is possible we could hide the manifests from users until we have more time to consider the functionality.

Reply:
We discussed in the arch call and enhancement has been updated